Show loading screen in urwid.

kaustubh-nair opened the pull request:
These commits are not perfectly isolated, because there are a lot of things to be done here:
* Show `LoadingView` first and run `Model` and `View` initialization asynchronously.
* Move settings from `run.py` to `LoadingView`
* Move spinning cursor from `Controller` to `LoadingView`

kaustubh-nair commented:
Making a new PR for this. 
I'll keep the current tutorial PR #583 only for tutorial related enhancements.

Adding tests.

sumanthvrao commented:
@kaustubh-nair I tried this out manually and it looks like great! 👌

neiljp commented:
@kaustubh-nair I replied on the stream about this - functionally this does what we want, but the structure is a little confusing and smaller commits may help explain which changes you're making and why. It's great that you're picking up others' older PRs and moving them forward, but it can be challenging precisely since they're not your code

Some aspects to consider:
* what (current) aspects of the model and view can be started in parallel or not?
* we have various cases in the code where we check whether the view is set up (`hasattr`); if we can ensure the ordering in the initial setup then can we remove these?
* while `Controller()` and its `.main(...)` are run consecutively, if we change the order we should understand/explain why
* when commits simply "move" code around, make sure that the code hasn't been adjusted by other commits between this - this is particularly true with rebasing older PRs!

kaustubh-nair committed:
refactor: core: Move Model and View initialization to new a method.

This is done so that it can be later run asynchronously by
introducing an intermediate view for loading.

kaustubh-nair committed:
core: Show loading screen in urwid.

Display LoadingView during startup. This blocks the settings shown in
stdout while loading which will be added in the subsequent commits.

kaustubh-nair committed:
core: views: Add spinning cursor to LoadingView.

kaustubh-nair committed:
core: Display settings in LoadingView.

Stop showing loading settings for theme and autohide  to stdout and use
the LoadingView instead. Note that in case of invalid settings,
output would still be displayed to stdout.

Tests amended.

sumanthvrao commented on zulipterminal/core.py:
```diff
@@ -273,16 +273,28 @@ def show_all_mentions(self, button: Any) -> None:
+                for cursor in '|/-\\':
+                    yield cursor
+
+        self.capture_stdout()
```
Do we need to capture stdout here again? We already do this now in the previous commit in the `__init__` stage itself.

sumanthvrao commented on zulipterminal/ui_tools/views.py:
```diff
@@ -1127,4 +1134,4 @@ def keypress(self, size: Tuple[int, int], key: str) -> str:
 
     def set_spinner(self, spinner: str) -> None:
-        self.base_widget.set_text(self.text + [spinner])
+        self.body.base_widget.set_text(self.text + [spinner])
```
Is this something necessary for this commit? Couldn't we move this to the previous commit where you introduce the `set_spinner` class?

sumanthvrao reviewed (commented):
@kaustubh-nair Thanks for rebasing 👍 . 

Additionally, I think this may avoid the UI distort race issue which we have seen in the past. I left a couple of commits but looks good overall.

Ezio-Sarthak commented on zulipterminal/core.py:
```diff
@@ -273,16 +273,28 @@ def show_all_mentions(self, button: Any) -> None:
     def deregister_client(self) -> None:
-        queue_id = self.model.queue_id
-        self.client.deregister(queue_id, 1.0)
+        # Deregister before model has loaded.
```
Is this necessary in this commit?
